dicompylercore/dose.py

```diff
@@ -0,0 +1,372 @@
+            self.z_axis = (
+                np.array(self.ds.GridFrameOffsetVector)
+                + self.ds.ImagePositionPatient[2]
+            )
+
+            # x and z are swapped in the pixel_array
+            pixel_array = self.ds.pixel_array * self.ds.DoseGridScaling
+            self.dose_grid = np.swapaxes(pixel_array, 0, 2)
+
+    ####################################################
+    # Basic properties
+    ####################################################
+    @property
+    def shape(self):
+        """Get the x, y, z dimensions of the dose grid"""
+        return tuple(
+            [self.ds.Columns, self.ds.Rows, len(self.ds.GridFrameOffsetVector)]
+        )
+
+    @property
+    def axes(self):
+        """Get the x, y, z axes of the dose grid (in mm)"""
+        return [self.x_axis, self.y_axis, self.z_axis]
+
+    @property
+    def scale(self):
+        """Get the dose grid resolution (xyz)"""
+        return np.array(
+            [
+                self.ds.PixelSpacing[0],
+                self.ds.PixelSpacing[1],
+                self.ds.GridFrameOffsetVector[1]
+                - self.ds.GridFrameOffsetVector[0],
+            ]
+        )
+
+    @property
+    def offset(self):
+        """Get the coordinates of the dose grid origin (mm)"""
+        return np.array(self.ds.ImagePositionPatient, dtype="float")
+
+    @property
+    def points(self):
+        """Get all of the points in the dose grid"""
+        y, x, z = np.meshgrid(self.y_axis, self.x_axis, self.z_axis)
+        points = np.vstack((x.ravel(), y.ravel(), z.ravel()))
+        return points.transpose()
+
+    ####################################################
+    # Tools
+    ####################################################
+    def __add__(self, other):
+        """Overload + operator to sum this dose grid with the other dose grid
+
+        Parameters
+        ----------
+        other : DoseGrid
+            Another DoseGrid object.
+        """
+        new = deepcopy(self)
+        new.add(other)
+        return new
+
+    def __mul__(self, factor):
+        """Overload * operator to scale this dose grid by the provided factor
+
+        Parameters
+        ----------
+        factor : int, float
+            Scale the dose grid by this value.
+        """
+        new = deepcopy(self)
+        new.multiply(factor)
+        return new
+
+    def __rmul__(self, factor):
+        return self.__mul__(factor)
+
+    def is_coincident(self, other):
+        """Check dose grid spatial coincidence.
+
+        Parameters
+        ----------
+        other : DoseGrid
+            Another DoseGrid object.
+        """
+        return (
+            self.ds.PixelSpacing == other.ds.PixelSpacing
+            and self.ds.ImagePositionPatient == other.ds.ImagePositionPatient
+            and self.ds.pixel_array.shape == other.ds.pixel_array.shape
+            and self.ds.GridFrameOffsetVector == other.ds.GridFrameOffsetVector
+        )
+
```

Comment:
Excellent feedback. Perhaps the way to go is to throw errors on these, but add a `force` flag?  Maybe still print the warnings if `force=True`?

I should have a commit to push this afternoon to address all of your comments.  Thanks!

Reply:
I think these are all addressed now.  Although, maybe it's more appropriate to make a function of the attr checks, and then call it in `DoseGrid.direct_sum` and `DoseGrid.interp_sum` instead of `DoseGrid.add`?

https://github.com/cutright/dicompyler-core/blob/a5bd38a4458411744f1398e1a644eb4c32d035c3/dicompylercore/dose.py#L229-L269

Reply:
Just reviewed again and these checks look good to me, although I think as you say, the attrs check should be done for the direct and interp paths, if the "API" allows people to call those directly.  Alternatively could make those "private" with a leading underscore.